File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int> left;
        int maxi = height[0];
        for(int i=0;i<height.size();i++){
            maxi = max(maxi,height[i]);
            left.push_back(maxi);
        }
        vector<int> right;
        maxi = height[height.size()-1];
        for(int i=height.size()-1;i>=0;i--){
            maxi = max(maxi,height[i]);
            right.push_back(maxi);
        }
        reverse(right.begin(),right.end());

        int ans =0;
        for(int i=0;i<height.size();i++){
            ans += min(left[i],right[i])-height[i];
        }
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left = []
        right = []
        maxi = 0
        for i in range(0,len(height)):
            maxi = max(maxi,height[i])
            left.append(maxi)
        maxi = 0
        for i in range(len(height)-1,-1,-1):
            maxi = max(maxi,height[i])
            right.append(maxi)
        right.reverse()
        ans =0
        for i in range(0,len(height)):
            ans += min(left[i],right[i])-height[i]
        return ans

        